### DISCUSSION TITLE
CSP: primefaces.nonce from request not validated

### DISCUSSION THREAD

**Discussion by danieldi96** [primefaces/community#2171] [PrimeFaces] [answered]
Hello,

Today I have created the follwing issue on Github<Primefaces. Github: https://github.com/primefaces/primefaces/issues/12140

I'm using JSF with Primefaces 8.0.25, spring-framework 5.3.34 and spring-security 5.8.12.
Primefaces creates a hidden input inside a form with a "primefaces.nonce" but using Tenable to test vulnerabilities, I get this error:

"
Identification
PROOF
The scanner submitted the form with a bad CSRF token value and got the same response from the web application.
OUTPUT
The scanner was able to detect that the form id:j_idt30 has a CSRF token named primefaces.nonce which is not
verified by the target web application.
"

I have a CSP vulnerability aswell:
"
OUTPUT
The scanner detected the following permissive statements in the detected effective Content Security Policy :
The 'default-src' directive has the following issue(s) :
- The directive is missing.
The 'object-src' directive has the following issue(s) :
- The directive is missing and the default-src directive is not set.
The 'base-uri' directive has the following issue(s) :
- The directive is missing.
The 'style-src' directive has the following issue(s) :
- The directive is missing and the default-src directive is not set.
The 'form-action' directive has the following issue(s) :
- The directive is missing."

On my web.xml, I have a CSP param which includes a 'script-src' header:

<context-param>
<param-name>primefaces.CSP</param-name>
<param-value>true</param-value>
</context-param>

I guess it is necessary to declare primefaces.CSP_POLICY, in order to declare the rest policies, should I?
Even though, I'm not sure how to declare nonce value inside CSP_POLICY because it changes every time (https://github.com/primefaces/primefaces/blob/master/docs/8_0/core/contentsecuritypolicy.md).

**Comment by melloware** [comment, marked as answer]
@danieldi96 you can customize the policy any way you want.

```js
<context-param>
    <param-name>primefaces.CSP_POLICY</param-name>
    <param-value>object-src 'self' script-src 'self' https://www.google-analytics.com</param-value>
</context-param>
```

For example PrimeFaces Extensions showcase uses this custom policy.

```xml
    <context-param>
        <param-name>primefaces.CSP_POLICY</param-name>
        <param-value>frame-ancestors 'none'; script-src 'self' blob: https: *.googleapis.com 'unsafe-hashes' 'sha256-kbHtQyYDQKz4SWMQ8OHVol3EC0t3tHEJFPCSwNG9NxQ='</param-value>
    </context-param>
```
